### PR TITLE
add an extra object to synthesized container for EnC

### PR DESCRIPTION
**Pull request by ivanbasov**
Towards #5156

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs [hunk 1]
```diff
@@ -123,16 +125,40 @@ internal override void AddSynthesizedAttributes(ModuleCompilationState compilati
...
+            }
+
             var ctor = Constructor;
             return ((object)ctor != null && name == ctor.Name) ? ImmutableArray.Create<Symbol>(ctor) : ImmutableArray<Symbol>.Empty;
         }
 
+        private Symbol GetStructFieldCaptureObject()
+        {
+            TypeSymbol objectType = this.DeclaringCompilation.GetSpecialType(SpecialType.System_Object);
+            return new SynthesizedFieldSymbol(this, objectType, StructFieldCaptureName, isPublic: false, isReadOnly: false, isStatic: false);
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: This method needs to be idempotent. It should never be possible to get two fields out of this method that are not reference equal to each other.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs [hunk 2]
```diff
@@ -15,6 +15,8 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
     /// </summary>
     internal abstract class SynthesizedContainer : NamedTypeSymbol
     {
+        internal const string StructFieldCaptureName = "<struct-field-capture-name>";
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: I think you need to come up with a real name for this before we merge it. Changing baselines because the name changed is annoying and we don't want to do it more than we have to.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs [hunk 3]
```diff
@@ -123,16 +125,40 @@ internal override void AddSynthesizedAttributes(ModuleCompilationState compilati
...
+            if ((object)constructor != null)
+            {
+                builder.Add(constructor);
+            }
+
+            return builder.ToImmutableAndFree();
         }
 
         public override ImmutableArray<Symbol> GetMembers(string name)
         {
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: Seems like this method should just call `GetMembers()` and filter. I don't know why it's doing what it's doing right now, but it seems weird.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs (hunk 2 above)
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: The field will need to be recognized by the EE to display the new local variables. As such it should be of the same format as other special generated names: ```<>``` {kind} -- see e.g. [http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Synthesized/GeneratedNames.cs,377](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Synthesized/GeneratedNames.cs,377)

The kind needs to be listed in [http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Synthesized/GeneratedNameKind.cs](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/Synthesized/GeneratedNameKind.cs) enum.

---
In reply to: [134590720](https://github.com/dotnet/roslyn/pull/21664#discussion_r134590720) [](ancestors = 134590720)

Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: I think ```v``` is the next kind that's not taken.

---
In reply to: [134593614](https://github.com/dotnet/roslyn/pull/21664#discussion_r134593614) [](ancestors = 134593614,134590720)

Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: >StructFieldCaptureName [](start = 30, length = 22)

I'd name this (and the generated name kind) EncStructExpansion

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs [hunk 4]
```diff
@@ -123,16 +125,40 @@ internal override void AddSynthesizedAttributes(ModuleCompilationState compilati
...
+            if (this.TypeKind == TypeKind.Struct && name == StructFieldCaptureName)
+            {
+                // Add an extra object field for EnC.
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: >// Add an extra object field for E [](start = 16, length = 34)

We should only add the field to synthesized containers that implement local function closures, not to async state machines. 

Also this field should only be added in debug builds.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs (hunk 2 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: thanks!

---
In reply to: [134594419](https://github.com/dotnet/roslyn/pull/21664#discussion_r134594419) [](ancestors = 134594419)

Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: Thanks!

---
In reply to: [134593971](https://github.com/dotnet/roslyn/pull/21664#discussion_r134593971) [](ancestors = 134593971,134593614,134590720)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs (hunk 3 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: Agree. Good idea! Thanks!

---
In reply to: [134591158](https://github.com/dotnet/roslyn/pull/21664#discussion_r134591158) [](ancestors = 134591158)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs (hunk 4 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: Thanks! Synced with Andy: moving this to SynthesizedClosureEnvironment

---
In reply to: [134595785](https://github.com/dotnet/roslyn/pull/21664#discussion_r134595785) [](ancestors = 134595785)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs (hunk 1 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedContainer.cs]: Thanks! Moved to the ctor

---
In reply to: [134590574](https://github.com/dotnet/roslyn/pull/21664#discussion_r134590574) [](ancestors = 134590574)

**Comment by ivanbasov** [comment]
test windows_debug_unit64_prtest please

**Comment by ivanbasov** [comment]
test windows_debug_unit64_prtest please

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureEnvironment.cs [hunk 5]
```diff
@@ -89,13 +98,20 @@ private static void AssertIsClosureScopeSyntax(SyntaxNode syntaxOpt)
 
         public override ImmutableArray<Symbol> GetMembers()
         {
-            var members = base.GetMembers();
+            var builder = ArrayBuilder<Symbol>.GetInstance();
+            builder.AddRange(base.GetMembers());
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureEnvironment.cs]: We might not want to change the order of members to avoid unnecessary changes in the emitted metadata.

I'd move AddRange to the end of the method.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs [hunk 6]
```diff
@@ -435,6 +435,12 @@ internal static string MakeIteratorCurrentThreadIdFieldName()
             return "<>l__initialThreadId";
         }
 
+        internal static string MakeEncStructExpansionFieldName()
+        {
+            Debug.Assert((char)GeneratedNameKind.EncStructExpansionField == 'v');
+            return "<>v_encStructExpansion";
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs]: The name pattern is double ```__```, but it may also be just ```<>v``` as that's unique already.

Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs]: Thank you! It seems <>v__encStructExpansion is the right choice.

---
In reply to: [134861840](https://github.com/dotnet/roslyn/pull/21664#discussion_r134861840) [](ancestors = 134861840)

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureEnvironment.cs (hunk 5 above)
Comment (ivanbasov) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureEnvironment.cs]: Sounds reasonable. Thank you!

---
In reply to: [134861344](https://github.com/dotnet/roslyn/pull/21664#discussion_r134861344) [](ancestors = 134861344)

**Comment by ivanbasov** [comment]
ping @tmat , @agocke , @dotnet/roslyn-compiler

**Comment by tmat** [comment]
I'm surprised that no test fail. Are we not testing shapes of local function closures?

**Comment by ivanbasov** [comment]
As I remember, we had multiple tests failing while we didn't check for ContainingSymbol.DeclaringCompilation.Options.EnableEditAndContinue.

File: src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureEnvironment.cs [hunk 7]
```diff
@@ -56,6 +57,13 @@ internal sealed class SynthesizedClosureEnvironment : SynthesizedContainer, ISyn
                 SingletonCache = new SynthesizedLambdaCacheFieldSymbol(this, this, cacheVariableName, topLevelMethod, isReadOnly: true, isStatic: true);
             }
 
+            if (isStruct && ContainingSymbol.DeclaringCompilation.Options.EnableEditAndContinue)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/SynthesizedClosureEnvironment.cs]: Can `ContainingSymbol` be `null` here?

**Comment by tmat** [comment]
@ivanbasov In that case we should duplicate those tests that were failing and compile them with TestOptions.Debug.

**Comment by ivanbasov** [comment]
Closing this PR for a while. We found that EnC support of variables captured by local functions is in parity with the support for lambdas. There is an area for improvements for both lambdas and local functions but we can enable EnC for local functions without this PR.
